apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java

```diff
@@ -182,9 +190,47 @@ public Span startSpan() {
         // user outcome to provide higher priority and avoid inferring outcome from any reported exception
         span.withUserOutcome(Outcome.UNKNOWN);
 
+        // Add the links to the span
+        // Presumably you'd only start the span once, so no need to
+        // worry about too much byte[] garbage ... but maybe we should
+        // pool across builders? Or maybe the OtelSpanBuilder should be recyclable?
+        byte[] header = new byte[TraceContext.BINARY_FORMAT_EXPECTED_LENGTH];
+        for (int i = 0; i < links.size(); i+=2) {
+            byte[] traceID = links.get(i);
+            byte[] spanID = links.get(i+1);
+            TraceContext.filIdOutgoingTraceParentBinaryHeader(header, traceID, spanID);
+            span.addSpanLink(
+                TraceContext.getFromTraceContextBinaryHeaders(),
+                BinaryByteArrayAccessor.INSTANCE,
+                header);
+        }
```

Comment:
I think this is forcing complicated logic (and allocation) only to match the current `AbstractSpan#addSpanLink` API, which assumes the caller is dealing with a header carrier and a dedicated accessor that knows how to get it based on name.
However, if you add other `AbstractSpan#addSpanLink()` APIs, that could get much simpler.

When we looked together, I thought you could add `AbstractSpan#addSpanLink(String/byte[] traceId, String/byte[] spanId)`, that could be, in which case the implementation would grab a `TraceContext` from the object pool and fill it with the required details (probably requiring a new `TraceContext#fromParent(String/byte[] traceId, String/byte[] parentId)`).

However, now that I looked at it again, it seems it may get even simpler than that - the `OTelSpanBuilder#addLink()` APIs receive a `io.opentelemetry.api.trace.SpanContext` as the parent. IIANM, this is always expected to be our implementation - `co.elastic.apm.agent.opentelemetry.sdk.OTelSpanContext`, which holds a `TraceContext` already.  So if you add an `AbstractSpan#addSpanLink(TraceContext parent)`, you're basically done. Since `co.elastic.apm.agent.opentelemetry.sdk.OTelSpanContext` seems immutable and non recyclable, you can add the `TraceContext` it refers to directly to the links list. Otherwise, if you fear it could be leaked, you can take a new `TraceContext` from the object pool and fill it through `copyFrom`.

On the other hand, I didn't really try anything, so orry if I am missing something....

Reply:
I just realized that since the span links are saved as children of the links, you cannot use the context itself. 
So the better API to add to `AbstractSpan` would be: 
```
public <T> boolean addSpanLink(ChildContextCreator<T> creator, T parent)
```
Of which implementation would still be extremely simple and very similar to the existing `addSpanLink`:
1. grab a `TraceContext` instance from the pool
2. initialize it through `creator.asChildOf()`

And to use it with a `TraceContext` you would do:
```
span.addSpanLink(TraceContext.fromParentContext(), parentTraceContext)
```

However, criticizing my own work- keeping the span links as children of the linked spans was a compromise of doing something less intuitive in order to avoid complicating `TraceContext`. However, I now think it makes sense to add the ability to create `TraceContext` from headers not as children and makes things more straightforward.
Let's talk about it...

Reply:
I've applied that advice (thanks!), have a new look